pkl-core/src/main/java/org/pkl/core/stdlib/yaml/ParserNodes.java

```diff
@@ -469,6 +473,13 @@ private void addMembers(MappingNode node, VmObject object) {
           var memberName =
               convertedKey instanceof String string && !useMapping ? Identifier.get(string) : null;
 
+          // https://github.com/apple/pkl/issues/561
+          if (memberName != null && "default".equals(convertedKey)) {
+            throw new ParseException(
```

Comment:
All other error handling code in this file throws `YamlEngineException`. Why throw a custom exception here and duplicate the code that catches `YamlEngineException` in the same file? (I know this was suggested in the review, but I don't see the benefit of adding all this code.)

Reply:
`YamlEngineException` doesn't feel like the right class to throw our _own_ errors with tailored messages. I think the direction we should move here is to switch all of our own throws from `YamlEngineException` to `org.pkl.core.util.yaml.ParseException`.

Reply:
Throwing ParseException everywhere is an improvement. But since these exceptions are implementation details and thrown and caught in the same file, introducing your own exception feels unnecessary (adds code for no benefit).

Same applies to json.ParserNodes: It throws org.pkl.core.util.json.ParseException, which also isn't your own exception.

Reply:
I agree that the extra class seems needlessly verbose and that we should probably move away from `YamlEngineException`s. I don't quite understand `org.pkl.core.util.json.ParseException` not being "your own exception." What do you mean by that, @translatenix?

Is there any reason not to move `org.pkl.core.util.json.ParseException` up one level (into `util`) and use it for either?

Reply:
The idea is that errors coming from `ParseException` can be presented to users as-is; because they are already tailored to be presented as error messages.

`YamlEngineException` messages come from SnakeYAML, and should be decorated with a generic "malformed yaml" message.

I think a good north star here is to not surface messages from `YamlEngineException` at all, and instead present all parse errors as our own tailored error messages. But, for now, it's helpful to have two different exception classes so we know what comes from us, and what comes from our parser library.

Reply:
> I don't quite understand org.pkl.core.util.json.ParseException not being "your own exception." What do you mean by that, @translatenix?

`org.pkl.core.util.json.ParseException` comes from the external JSON parser library that was copied into Pkl's codebase. As such it's the equivalent of `YamlEngineException` .

> YamlEngineException messages come from SnakeYAML, and should be decorated with a generic "malformed yaml" message.

If that's what you want, having your own exception makes sense, assuming you can't throw a `VmException` to begin with. You may also want to make this change for the JSON parser.